### ISSUE TITLE
`Conditional` decorator

### ISSUE THREAD

**Issue by dmfs** [dmfs/optional#8]
As the counterpart to the `Filtered` `Iterator` decorator, it could be useful to have a `Conditional` `Optional` which is only present if the decorated `Optional` is present *and* the additional `Filter` returns `true`.
@lemonboston what do you think?

Not sure if `Conditional` is the correct name here, because it doesn't make an absent `Optional` present, just because the condition is met. So maybe `Filtered` suits better in this case as well.

**Comment by lemonboston** [comment]
Created PR with the name `Filtered`... more comments there.

**Comment by dmfs** [comment]
Merged.

Regarding the `Predicate` vs. `Filter` issue, I was considering that. Actually I have a branch in which I renamed `Filter` to `Predicate` to match the Java 8 terminology.

However, I see a problem with that name. We currently have a mutable `Filter` (which is `Unique`). To me it seems to be fine to have "Filter" with a mutable state (just like an electronic filter does have a mutable state, implemented by capacities and coils). But I can't imagine a "Predicate" with a mutable state (to me a predicate is something that does not change).

Also I don't like the method name `test` of the Java 8 `Predicate` interface.
